### ISSUE TITLE
SparseSoftmaxCrossEntropyWithLogits 계산시 Label 개수 에러

### ISSUE THREAD

**Issue by bart2001** [nlp-research/biaffineparser#30]
```python
Caused by op 'SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits', defined at:
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/nhnent/www/biaffineparser/parser/parser.py", line 221, in <module>
    print('Done')
  File "/Users/nhnent/www/biaffineparser/parser/parser.py", line 145, in main
    model = Model(flags, words_dict, pos_features_dict, rels_features_dict, heads_features_dict, word_embedding, pos_embedding)
  File "/Users/nhnent/www/biaffineparser/parser/model.py", line 40, in __init__
    self.build()
  File "/Users/nhnent/www/biaffineparser/parser/model.py", line 64, in build
    self.create_loss_op()
  File "/Users/nhnent/www/biaffineparser/parser/model.py", line 188, in create_loss_op
    self.arc_logits, self.head_ids, self.sequence_length)
  File "/Users/nhnent/www/biaffineparser/parser/model.py", line 177, in compute_loss
    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=gold_labels)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/site-packages/tensorflow/python/ops/nn_ops.py", line 2085, in sparse_softmax_cross_entropy_with_logits
    precise_logits, labels, name=name)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 7519, in sparse_softmax_cross_entropy_with_logits
    labels=labels, name=name)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 454, in new_func
    return func(*args, **kwargs)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3155, in create_op
    op_def=op_def)
  File "/Users/nhnent/miniconda3/envs/parser/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1717, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): Received a label value of 51 which is outside the valid range of [0, 51)
```
train step에서 heads_indexed_batch를 통해서 cross_entropy 계산시 위와 같은 에러가 발생합니다.
head_feature_dic에서 51은 "<PAD>"에 해당합니다.

heads_indexed_batch의 경우 (sejong_test_set 기준) 다음과 같은 형태를 띄고 있습니다.
```python
heads_indexed_batch[1] = array([ 0,  1,  4, 17,  4, 18, 18, 23,  6, 23, 21,  5, 23, 23,  8,  7, 24,
       24,  0, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51,
       51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51])
print(len(heads_indexed_batch[1])) # 51
```
51이 PAD값인데 51이 들어가 있는 것이 왜 문제인지 파악중입니다.

**Comment by bart2001** [comment]
head_vocab(head_features)의 개수가 최대 seqence_length를 초과하여 생기는 현상입니다.

예를 들어 1개의 sentence의 head_id를 표현할 때 [1,2,3,4,5,6,7,8,9,0]이 되어야 하는데 [1,2,3,4,5,6,7,8,9,10]과 같이 전체의 길이 10을 넘는 데이터 10이 존재하기 때문입니다.

* 위의 현상은 tf.nn.sparse_softmax_cross_entropy_with_logits 계산시에 오류를 발생합니다.

최대 sequence_length(maxlen)이 head_features의 개수보다 적을 경우 강제로 sequence_length(maxlen)를 업데이트 해주는 방식으로 처리하였습니다.

```python
def load_dataset(filepath):
    sentences, pos, rels, heads, maxlen, maxwordlen = get_dataset_multiindex(filepath)
    _, heads_features_dict, _ = initialize_embed_features(heads, 100, maxlen, starti=0, return_embeddings=False)
    #update maxlen when maxlen is less or equal to number of head_features (head_vocab)
    if maxlen < len(heads_features_dict):
        print(f'updating maxlen since it is less than head_features :maxlen={maxlen}')
        maxlen = len(heads_features_dict)
```
